tests/util/test_linearizer.py

```diff
@@ -13,160 +13,205 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-from twisted.internet import defer, reactor
-from twisted.internet.defer import CancelledError
+from typing import Callable, Hashable, Optional, Tuple
 
-from synapse.logging.context import LoggingContext, current_context
-from synapse.util import Clock
+from twisted.internet import defer, reactor
+from twisted.internet.base import ReactorBase
+from twisted.internet.defer import CancelledError, Deferred
+
+from synapse.logging.context import (
+    LoggingContext,
+    current_context,
+    make_deferred_yieldable,
+)
 from synapse.util.async_helpers import Linearizer
 
 from tests import unittest
 
 
 class LinearizerTestCase(unittest.TestCase):
-    @defer.inlineCallbacks
-    def test_linearizer(self):
+    def _start_task(
+        self, linearizer: Linearizer, key: Hashable
+    ) -> Tuple["Deferred[None]", "Deferred[None]", Callable[[], None]]:
+        """Starts a task which acquires the linearizer lock, blocks, then completes.
+
+        Args:
+            linearizer: The `Linearizer`.
+            key: The `Linearizer` key.
+
+        Returns:
+            A tuple containing:
+             * A cancellable `Deferred` for the entire task.
+             * A `Deferred` that resolves once the task acquires the lock.
+             * A function that unblocks the task. Must be called by the caller
+               to allow the task to release the lock and complete.
+        """
+        acquired_d: "Deferred[None]" = Deferred()
+        unblock_d: "Deferred[None]" = Deferred()
+
+        async def task() -> None:
+            with await linearizer.queue(key):
+                acquired_d.callback(None)
```

Comment:
Unlike the previous code, we enter the context manager immediately once `linearizer.queue()` resolves.

So
```py
cm1 = yield d1
with cm1:
    self.assert(...)
```
turns into
```py
# context manager entered already
self.assertTrue(acquired_d1.called)

self.assert(...)

unblock()  # exit the context manager
```

And the asserts that were once inside the context manager often become duplicates and are removed.